File: app/src/main/java/in/completecourse/adapter/CarouselPagerAdapter.java
package in.completecourse.adapter;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import in.completecourse.R;
import in.completecourse.SubjectActivity;
import in.completecourse.fragment.ItemFragment;
import in.completecourse.utils.CarouselLinearLayout;
import in.completecourse.utils.ListConfig;

public class CarouselPagerAdapter extends FragmentPagerAdapter implements ViewPager.OnPageChangeListener {

    private final static float BIG_SCALE = 1.0f;
    private final static float SMALL_SCALE = 0.7f;
    private final static float DIFF_SCALE = BIG_SCALE - SMALL_SCALE;
    private final SubjectActivity context;
    private final FragmentManager fragmentManager;
    private float scale;

    public CarouselPagerAdapter(SubjectActivity context, FragmentManager fm) {
        super(fm);
        this.fragmentManager = fm;
        this.context = context;
    }

    @NonNull
    @Override
    public Fragment getItem(int position) {
        // make the first pager bigger than others
        try {
            if (position == SubjectActivity.FIRST_PAGE)
                scale = BIG_SCALE;
            else
                scale = SMALL_SCALE;
            if (SubjectActivity.classString.equalsIgnoreCase("4") ||
                    SubjectActivity.classString.equalsIgnoreCase("1")) {
                position = position % ListConfig.subjectHighSchool.length;
            }else{
                position = position % SubjectActivity.count;
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
        return ItemFragment.newInstance(context, position, scale);
    }

    @Override
    public int getCount() {
        int count = 0;
        try {
            count = SubjectActivity.count * SubjectActivity.LOOPS;
        } catch (Exception e) {
            // TODO: handle exception
            e.printStackTrace();
        }
        return count;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        try {
            if (positionOffset >= 0f && positionOffset <= 1f) {
                CarouselLinearLayout cur = getRootView(position);
                CarouselLinearLayout next = getRootView(position + 1);
                cur.setScaleBoth(BIG_SCALE - (DIFF_SCALE * positionOffset));
                next.setScaleBoth(SMALL_SCALE + (DIFF_SCALE * positionOffset));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    @SuppressWarnings("ConstantConditions")
    private CarouselLinearLayout getRootView(int position) {
        return (CarouselLinearLayout) fragmentManager.findFragmentByTag(this.getFragmentTag(position))
                .getView().findViewById(R.id.root_container);
    }

    private String getFragmentTag(int position) {
        return "android:switcher:" + SubjectActivity.subjectViewPager.getId() + ":" + position;
    }
}

File: app/src/main/java/in/completecourse/adapter/CarouselPagerAdapter.kt
package `in`.completecourse.adapter

import `in`.completecourse.R
import `in`.completecourse.SubjectActivity
import `in`.completecourse.fragment.ItemFragment
import `in`.completecourse.utils.CarouselLinearLayout
import `in`.completecourse.utils.ListConfig
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import kotlinx.android.synthetic.main.activity_subject.*

class CarouselPagerAdapter(private val context: SubjectActivity,
                           private val fragmentManager: FragmentManager,
                           private val classString:String?,
                           private val subjectString:String? ) : FragmentPagerAdapter(fragmentManager), OnPageChangeListener {
    private var scale = 0f

    override fun getItem(position: Int): Fragment {
        // make the first pager bigger than others
        var position = position
        try {
            scale = if (position == SubjectActivity.FIRST_PAGE) BIG_SCALE else SMALL_SCALE
            position %= if (classString.equals("4", ignoreCase = true) ||
                    classString.equals("1", ignoreCase = true)) {
                ListConfig.subjectHighSchool.size
            } else {
                SubjectActivity.count
            }
        }
        catch (e: Exception) { e.printStackTrace() }
        return ItemFragment.newInstance(context, position, scale, classString, subjectString)
    }

    override fun getCount(): Int {
        var count = 0
        try {
            count = SubjectActivity.count * SubjectActivity.LOOPS
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return count
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        try {
            if (positionOffset in 0f..1f) {
                val cur: CarouselLinearLayout = getRootView(position)
                val next: CarouselLinearLayout = getRootView(position + 1)
                cur.setScaleBoth(BIG_SCALE - DIFF_SCALE * positionOffset)
                next.setScaleBoth(SMALL_SCALE + DIFF_SCALE * positionOffset)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onPageSelected(position: Int) {}
    override fun onPageScrollStateChanged(state: Int) {}
    private fun getRootView(position: Int): CarouselLinearLayout {
        return fragmentManager.findFragmentByTag(getFragmentTag(position))
                ?.view!!.findViewById<View>(R.id.root_container) as CarouselLinearLayout
    }

    private fun getFragmentTag(position: Int): String {
        return "android:switcher:" + context.viewpagerr.id + ":" + position
    }

    companion object {
        private const val BIG_SCALE = 1.0f
        private const val SMALL_SCALE = 0.7f
        private const val DIFF_SCALE = BIG_SCALE - SMALL_SCALE
    }

}